lib/vector-core/src/event/metric/data.rs

```diff
@@ -49,36 +60,50 @@ impl MetricData {
     #[must_use]
     pub fn into_incremental(self) -> Self {
         Self {
-            timestamp: self.timestamp,
+            time: self.time,
             kind: MetricKind::Incremental,
             value: self.value,
         }
     }
 
     /// Creates a `MetricData` directly from the raw components of another `MetricData`.
-    pub fn from_parts(
-        timestamp: Option<DateTime<Utc>>,
-        kind: MetricKind,
-        value: MetricValue,
-    ) -> Self {
-        Self {
-            timestamp,
-            kind,
-            value,
-        }
+    pub fn from_parts(time: MetricTime, kind: MetricKind, value: MetricValue) -> Self {
+        Self { time, kind, value }
     }
 
     /// Decomposes a `MetricData` into its individual parts.
-    pub fn into_parts(self) -> (Option<DateTime<Utc>>, MetricKind, MetricValue) {
-        (self.timestamp, self.kind, self.value)
+    pub fn into_parts(self) -> (MetricTime, MetricKind, MetricValue) {
+        (self.time, self.kind, self.value)
     }
 
     /// Updates this metric by adding the value from `other`.
     #[must_use]
     pub fn update(&mut self, other: &Self) -> bool {
         self.value.add(&other.value) && {
+            let overlap = if let (Some(t1), Some(i1), Some(t2), Some(i2)) = (
+                self.time.timestamp,
+                self.time.interval_ms,
+                other.time.timestamp,
+                other.time.interval_ms,
+            ) {
+                if t1.timestamp_millis() > t2.timestamp_millis() {
+                    t1.timestamp_millis() <= t2.timestamp_millis() + (i2.get() as i64)
+                } else {
+                    t2.timestamp_millis() <= t1.timestamp_millis() + (i1.get() as i64)
+                }
+            } else {
+                false
+            };
+            // Update the interval covered by the resulting metric
+            self.time.interval_ms = match (self.time.interval_ms, other.time.interval_ms, overlap) {
+                // If metrics time span are overlapping (should be the usual case) we use the longest interval
+                (Some(i1), Some(i2), true) => NonZeroU32::new(i1.get().max(i2.get())),
+                (Some(i1), Some(i2), false) => NonZeroU32::new(i1.get() + i2.get()),
+                // If either interval is None discard the other
+                (_, _, _) => None,
+            };
```

Comment:
Hmmm, I haven't fully thought through the logic here, but I'm worried this will lead to incorrect results in terms of the resulting rate (value / interval) of the updated metric.

I think the logic for discarding the interval when they both don't have one is correct, but what worries me is our calculation when intervals are present. I would intuitively expect to calculate the new interval with something like:

```rust
let (new_ts, new_interval) = if t1 > t2 {
  // The interval window starts from the beginning of `other` (aka `t2`)
  // and goes to the end of `self` (which is `t1 + i1`).
  (t2, (t1 - t2) + i1)
} else {
  // The interval window starts from the beginning of `self` (aka `t1`)
  // and goes to the end of `other` (which is `t2 + i2`).
  (t1, (t2 - t1) + i2)
};
```

### When `self` is newer than `other`, no overlap

Concretely, if t1=50, i1=15, t2=15, i2=10, the sum of the those two metrics represents values over a total interval of 50 (`50 - 15 + 15`), starting at 15 (`t2`).

If the values added were 30 and 30 for `self` and `other`, respectively, we would expect their individual rates to be 2 and 3, respectively. Added together, this now becomes a value of 60 over an interval of 50, or a rate of 1.2.

With the code as-is in the PR, the timestamp becomes the higher of the two, which is 50, and the interval becomes 15 + 10, or 25. This results in a rate of 2.4, which is close to the average of the rates of the two individual metrics... but since the metrics don't overlap, we can't simply average them because it will generate a metric that shifts forward in the future and doesn't correctly demonstrate that the values were incremented in the time window between/including the two metrics, rather than just timeshifted up.

### When `self` is older than `other`, no overlap

Concretely, if t1=35, i1=5, t2=50, i2=10, the sum of the those two metrics represents values over a total interval of 25 (`50 - 35 + 10`), starting at 35 (`t1`).

If the values added were 50 and 50 for `self` and `other`, respectively, we would expect their individual rates to be 10 and 5, respectively. Added together, this now becomes a value of 100 over an interval of 25, or a rate of 4.

With the code as-is in the PR, the timestamp becomes the higher of the two, which is 50, and the interval becomes 5 + 10, or 15. This results in a rate of 6.6666..., which is somewhat close to the average of the rates of the two individual metrics... but since the metrics don't overlap, we can't simply average them because it will generate a metric that shifts backwards in time and doesn't correctly demonstrate that the values were incremented in the time window between/including the two metrics, rather than just timeshifted down.

While we've talked about what happens above when there is no overlap, in the overlap case, we're still incorrectly (IMO) calculating our new timestamp/interval. This one is tricky because there shouldn't really be a case where a single upstream system can change the interval and still have overlapping windows, although it's technically possible. More likely, this would bite us during aggregation of metrics from different systems which _could_ have disparate intervals.

As an extreme example, let's assume t1=5 i1=10 v1=2000 and t2=10 i2=10 v2=25, which gives us the two resulting rates, r1=200 and r2=2.5.

The true interval for this scenario would be 15, as the time window would start at 5 (`t1`) and go to 20 (`t2 + i2`). Added together, the total value would be 2025, for an averaged rate of 135. With the code as-is in the PR, we would take the larger of the two intervals (10, both `i1` and `i2`) and the more recent of the two timestamps (10, `t2`) as our new interval and timestamp, respectively.

Using the new timestamp of 10 and interval of 10, not only would the rate jump to 202.5 (2025 / 10) but we would be shifting those values forward in time by 5, which would misrepresent the large spike in v1 by timeshifting it forward.